qutip/solver/mcsolve.py

```diff
@@ -0,0 +1,311 @@
+    """
+    Solver for one mcsolve trajectory. Created by a :class:`McSolver`.
+    """
+    name = "mcsolve"
+    _avail_integrators = {}
+
+    def __init__(self, rhs, c_ops, n_ops, *, options=None, _prepare=False):
+        self._c_ops = c_ops
+        self._n_ops = n_ops
+        super().__init__(rhs, options=options, _prepare=_prepare)
+
+    def _run(self, seed, state, tlist, e_ops):
+        self.collapses = []
+        self.generator = self.get_generator(seed)
+        self.target_norm = self.generator.random()
+        result = super()._run(self.generator, state, tlist, e_ops)
+        result.collapse = list(self.collapses)
+        result.seed = seed
+        return result
+
+    def start(self, state, t0, seed=None, *, safe_ODE=None):
+        self.collapses = []
+        self.generator = self.get_generator(seed)
+        self.target_norm = self.generator.random()
+        super().start(state, t0, seed=self.generator, safe_ODE=safe_ODE)
+
+    def prob_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)**2
+
+    def norm_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)
+
+    def _step(self, t, copy=True):
+        t_old, y_old = self._integrator.get_state(copy=False)
+        norm_old = self.prob_func(y_old)
+        while t_old < t:
+            t_step, state = self._integrator.mcstep(t, copy=False)
+            norm = self.prob_func(state)
+            if norm <= self.target_norm:
+                self._do_collapse(norm_old, norm, t_old)
+                t_old, y_old = self._integrator.get_state(copy=False)
+                norm_old = 1.
+            else:
+                t_old, y_old = t_step, state
+                norm_old = norm
+
+        return t_old, _data.mul(y_old, 1 / self.norm_func(y_old))
+
+    def _do_collapse(self, norm_old, norm, t_prev):
+        """Find and apply a collapse."""
+        t_final, _ = self._integrator.get_state()
+        tries = 0
+        while tries < self.options.mcsolve['norm_steps']:
+            tries += 1
+            if (t_final - t_prev) < self.options.mcsolve['norm_t_tol']:
+                t_guess = t_final
+                _, state = self._integrator.get_state()
+                break
+            t_guess = (
+                t_prev
+                + ((t_final - t_prev)
+                   * np.log(norm_old / self.target_norm)
+                   / np.log(norm_old / norm))
+            )
+            if (t_guess - t_prev) < self.options.mcsolve['norm_t_tol']:
+                t_guess = t_prev + self.options.mcsolve['norm_t_tol']
+            _, state = self._integrator.mcstep(t_guess, copy=False)
+            norm2_guess = self.prob_func(state)
+            if (
+                np.abs(self.target_norm - norm2_guess) <
+                self.options.mcsolve['norm_tol'] * self.target_norm
```

Comment:
I see that `norm_tol` refers to relative, unlike `norm_t_tol`. we need to be clear on this in the documentation. I would suggest a slight renaming to `norm_rtol` and `norm_t_atol` so that it is already clear from the name what kin of tolerances each refer to.

Reply:
The atol vs rtol does not matter enough to change the names but documentation not mentioning that it's used when finding collapse is certainly lacking.